File: Sources/SomnioCore/Models/InventoryRow.swift
import Foundation

/// Ordered representation of category-specific extras (e.g. `{"gold": N}` for the purse).
/// The ordered-array form replaces a raw `[String: Int16]` so binary serialization is
/// deterministic; runtime code that needs map-style lookup constructs a transient
/// `Dictionary(uniqueKeysWithValues: row.extras.map { ($0.key, $0.value) })`.
public struct InventoryExtra: Codable, Sendable, Equatable, Hashable {
    /// Extra key carrying the purse's coin balance. The wire key is a property name, so the literal
    /// is shared here rather than written at each producer/consumer (a typo would silently read 0).
    public static let goldKey = "gold"

    public var key: String
    public var value: Int16

    public init(key: String, value: Int16) {
        self.key = key
        self.value = value
    }
}

public struct InventoryRow: Sendable, Equatable, Hashable {
    public var slot: Int16
    public var category: Int16
    public var itemId: Int16
    public var extras: [InventoryExtra]
    public var equippedHand: Hand?

    public init(
        slot: Int16,
        category: Int16,
        itemId: Int16,
        extras: [InventoryExtra] = [],
        equippedHand: Hand? = nil
    ) {
        self.slot = slot
        self.category = category
        self.itemId = itemId
        self.extras = extras
        self.equippedHand = equippedHand
    }

    /// Coin balance carried by the `gold` extra (the purse), or 0 when absent.
    public var goldBalance: Int16 {
        extras.first { $0.key == InventoryExtra.goldKey }?.value ?? 0
    }
}

File: packages/core/src/domain/inventoryRow.ts
/**
 * The hand an item is equipped in. Distinct from the wire's `WIRE_HAND` by one: the wire reserves
 * 0 for "no hand" while unequipped is `undefined` here, so the two must never be compared
 * directly — `handFromWire`/`handToWire` are the conversions.
 */
export const HAND = { left: 0, right: 1 } as const;
export type Hand = (typeof HAND)[keyof typeof HAND];

/** Extra key carrying the purse's coin balance. A typo here would silently read 0. */
export const GOLD_KEY = 'gold';

/**
 * Ordered category-specific extras (`{"gold": N}` for the purse). The ordered-array form keeps
 * serialization deterministic; map-style lookup builds a transient index.
 */
export interface InventoryExtra {
  key: string;
  value: number;
}

export interface InventoryRow {
  slot: number;
  category: number;
  itemId: number;
  extras: InventoryExtra[];
  /** `undefined` when unequipped; see `HAND` for the two equipped values. */
  equippedHand: Hand | undefined;
}

/** Coin balance carried by the `gold` extra (the purse), or 0 when absent. */
export function goldBalance(row: InventoryRow): number {
  return row.extras.find((extra) => extra.key === GOLD_KEY)?.value ?? 0;
}
